src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -3,122 +3,162 @@
+                var epoch = GetEpoch(now.Ticks);
+                var howMany = intervalsToFill.Length;
 
-                if (timings.TryGetValue(instanceId, out endpointInstanceData))
-                {
-                    var totalDuration = 0L;
-                    var totalMeasurements = 0L;
-                    var timingIntervals = new TimingInterval[intervals.Length];
+                totalDuration = 0;
+                totalMeasurements = 0;
 
-                    for (var i = 0; i < intervals.Length; i++)
+                rwl.EnterReadLock();
+                try
+                {
+                    for (var i = 0; i < howMany; i++)
                     {
-                        timingIntervals[i] = new TimingInterval
+                        var epochIndex = epoch % Size;
+                        var interval = intervals[epochIndex];
+
+                        intervalsToFill[i] = new TimingInterval
                         {
-                            IntervalStart = intervals[i]
+                            IntervalStart = GetDateTime(epoch),
                         };
 
-                        MeasurementInterval bucket;
-
-                        if (endpointInstanceData.TryGetValue(intervals[i], out bucket))
+                        // the interval might contain data from the right epoch, or epochs before that have the same index
+                        // we calculate data only if that's the right epoch
+                        if (interval.Epoch == epoch)
                         {
-                            totalDuration += bucket.TotalTime;
-                            totalMeasurements += bucket.TotalMeasurements;
+                            intervalsToFill[i].TotalTime = interval.TotalTime;
+                            intervalsToFill[i].TotalMeasurements = interval.TotalMeasurements;
 
-                            timingIntervals[i].TotalTime = bucket.TotalTime;
-                            timingIntervals[i].TotalMeasurements = bucket.TotalMeasurements;
+                            totalDuration += interval.TotalTime;
+                            totalMeasurements += interval.TotalMeasurements;
                         }
+
+                        epoch -= 1;
                     }
+                }
+                finally
+                {
+                    rwl.ExitReadLock();
+                }
+            }
 
-                    result.Add(new EndpointInstanceTimings
+            public void Report(LongValueOccurrences message)
+            {
+                rwl.EnterWriteLock();
+                try
+                {
+                    for (var i = 0; i < message.Length; i++)
                     {
-                        Id = instanceId,
-                        TotalMeasurements = totalMeasurements,
-                        TotalTime = totalDuration,
-                        Intervals = timingIntervals
-                    });
+                        Report(ref message.entries[i]);
```

Comment:
Entries are `structs` to make them allocated as a single table. I'm using `ref` to pass a "reference to a struct in a table".